### PR TITLE
1.3/1068 add one or more vouchers to a bundle backend

### PR DESCRIPTION
**Pull request by charlesstrange2**
Ok! Should accept and guess a bunch of vouchers based on start and end!

### CODE REVIEW

File: app/Bundle.php [hunk 1]
```diff
@@ -72,27 +107,28 @@ class Bundle extends Model
     public function alterVouchers(Collection $vouchers, array $codes = [], Bundle $bundle = null)
```
Comment (katjam) [inline app/Bundle.php]: I'm not that keen on the $self/ $this reference of Bundle. Might simplify things if addVouchers took care of this... but if you are using to replace whole bundle elsewhere we need to keep?

File: app/Http/Controllers/Store/BundleController.php [hunk 2]
```diff
@@ -69,19 +128,7 @@ public function update(StoreUpdateBundleRequest $request, Registration $registra
...
-                ->with('error_message', ucfirst(join(', ', $messages) . '.'));
-        } else {
-            // Otherwise, sure, return to the new view.
-            return redirect()->route('store.registration.voucher-manager', ['registration' => $registration->id])
-                ->with('message', 'Vouchers updated.');
-        }
+        return $this->redirectAfterRequest($bundle->syncVouchers($voucherCodes), $registration);
     }
 
     public function removeVoucherFromCurrentBundle(Registration $registration, Voucher $voucher)
```
Comment (katjam) [inline app/Http/Controllers/Store/BundleController.php]: Missing doc

File: app/Http/Controllers/Store/BundleController.php [hunk 3]
```diff
@@ -91,54 +138,57 @@ public function removeVoucherFromCurrentBundle(Registration $registration, Vouch
...
             $errors = $bundle->alterVouchers(collect([$voucher]), [], null);
         } else {
             // Error it out (how did you get here?
             $errors["foreign"] = [$voucher->code];
         }
 
+        return $this->redirectAfterRequest($errors, $registration);
+    }
+
+    public function redirectAfterRequest($errors, $registration)
```
Comment (katjam) [inline app/Http/Controllers/Store/BundleController.php]: Missing doc

File: app/Voucher.php [hunk 4]
```diff
@@ -81,6 +80,27 @@ function ($code) {
...
+     */
+    public static function splitShortcodeNumeric(string $code)
+    {
+        // Clean the code
+        $clean = self::cleanCodes([$code]);
+        $code = array_shift($clean);
+        // Init matches
+        $matches = [];
+        // split into named matche and return
+        if (preg_match("/^(?<shortcode>\D*)(?<number>\d+)$/", $code, $matches) == 1) {
```
Comment (katjam) [inline app/Voucher.php]: Assuming this regex works.

File: app/Http/Controllers/Store/BundleController.php [hunk 5]
```diff
@@ -51,7 +52,65 @@ public function index()
...
     /**
-     * Create OR Update a registrations current active bundle
+     * Does a single voucher.
+     *
+     * @param StoreAppendBundleRequest $request
+     * @param Registration $registration
+     * @return \Illuminate\Http\RedirectResponse
+     */
+
+    public function addVouchersToCurrentBundle(StoreAppendBundleRequest $request, Registration $registration)
```
Comment (katjam) [inline app/Http/Controllers/Store/BundleController.php]: Consider sharing some functionality with `/app/Http/Controllers/Service/Admin/VouchersController.php`

The diff is that submit comes with Sponsor id and numbers - but might be able to use some rules and `range()`

File: app/Voucher.php [hunk 6]
```diff
@@ -202,17 +222,18 @@ public static function findByCodes($codes)
    *
    * @return array
    */
-  public static function getMinMaxVoucherDates(Collection $vouchers) {
```
Comment (katjam) [inline app/Voucher.php]: Nothing new here, right?

File: app/Voucher.php (hunk 4 above)
Comment (charlesstrange2) [inline app/Voucher.php]: I did some manual testing on it. It uses fancy named group return values. Coild do with an auto test. Ill write one seperatly

File: app/Voucher.php (hunk 6 above)
Comment (charlesstrange2) [reply app/Voucher.php]: Nope, its just a bit of formatting

File: app/Bundle.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Bundle.php]: I'm not either, but it happened organically while i was avoiding writing the same thing everywhere

**Pull request merged**